apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4SamhsaMatcher.java

```diff
@@ -522,43 +523,89 @@ private static String normalizeIcd10Code(String icd10DiagnosisCode) {
   /**
    * @param procedureConcept the procedure {@link CodeableConcept} to check
    * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains any
-   *     {@link Coding}s that match any of the {@link #cptCodes}, <code>false</code> if they all do
-   *     not
+   *     {@link Coding}s that match any of the {@link #cptCodes} and has no unknown {@link System}s,
+   *     <code>false</code> if they all do not
    */
   private boolean containsSamhsaProcedureCode(CodeableConcept procedureConcept) {
+    // Does the CodeableConcept have a legit HCPCS Coding?
+    boolean hasHcpcsCoding = findHcpcsCoding(procedureConcept);
+
+    // Check that Coding to see if it's blacklisted.
+    if (hasHcpcsCoding && isSamhsaCptCode(procedureConcept)) {
+      return true;
+    } else if (hasHcpcsCoding && hasUnknownSystem(procedureConcept)) {
+      /*
+       * Fail safe: if we don't know the procedure Coding system, assume the code is
+       * SAMHSA.
+       */
+      return true;
+    } else {
+      // Otherwise, this only has known & non-SAMHSA-blacklisted procedure codes.
+      return false;
+    }
+  }
+
+  /**
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains the
+   *     {@link Coding} with the HCPCS {@link System}, <code>false</code> if it does not
+   */
+  private boolean findHcpcsCoding(CodeableConcept procedureConcept) {
```

Comment:
Might be better called "hasHcpcsCoding" or "containsHcpcsCoding" since we're not retrieving the Coding here and we're returning a boolean.